src/resources/action.ts

```diff
@@ -0,0 +1,141 @@
+import { Macro } from "../combat";
+import { Requirement } from "../maximize";
+
+export type FindActionSourceOptions = {
+  requireFree?: () => boolean;
+  requireFamiliar?: () => boolean;
+  requireUnlimited?: () => boolean;
+  noFamiliar?: () => boolean;
+  noRequirements?: () => boolean;
+  noPreparation?: () => boolean;
+  maximumCost?: () => number;
+};
+
+export type ActionOptions = {
+  equipmentRequirements?: () => Requirement;
+  preparation?: () => boolean;
+  familiar?: () => Familiar;
+  cost?: () => number;
+};
+
+function mergeOptions(...options: ActionOptions[]): ActionOptions {
+  return {
+    equipmentRequirements: () =>
+      Requirement.merge([
+        ...options.map(
+          (option) =>
+            option.equipmentRequirements?.() ?? new Requirement([], {})
+        ),
+      ]),
+    preparation: () => {
+      let success = true;
+      options.forEach(
+        (option) =>
+          (success = success && (!option?.preparation || option?.preparation()))
+      );
+      return success;
+    },
+    // Return thet firs familiar if multiples need them
+    familiar: options
+      .map((option) => option.familiar)
+      .find((familiar) => familiar),
+    cost: () =>
+      options
+        .map((option) => option.cost?.() ?? 0)
+        .reduce((previous, current) => previous + current),
+  };
+}
+
+export class ActionSource {
+  source: Item | Skill | Familiar;
+  potential: () => number; // -1: unlimited
+  macro: Macro;
+  options?: ActionOptions;
+
+  constructor(
+    source: Item | Skill | Familiar,
+    potential: () => number,
+    macro: Macro,
+    options?: ActionOptions
+  ) {
+    this.source = source;
+    this.potential = potential;
+    this.macro = macro;
+    this.options = options;
+  }
+
+  name(): string {
+    return `${this.source}`;
+  }
+
+  available(): boolean {
+    return this.potential() !== 0;
```

Comment:
```suggestion
    return this.potential() > 0;
```

This way we don't have to be as careful about making sure things don't go negative (and more robust to tracking bugs, I think).